### PR TITLE
UUID serializer and its unit test

### PR DESCRIPTION
**Pull request by BishrutSubedi**
This PR adds UUID serializer for uprotocol uuid v8  #122

### CODE REVIEW

File: src/datamodel/serializer/Uuid.cpp
Comment (gregmedd) [inline src/datamodel/serializer/Uuid.cpp:24]: Changing this to `return std::move(ss).str();` tells the compiler that `ss` is no longer needed, allowing the returned string to be move constructed. This saves an allocate, copy, deallocate cycle (so long as the string is longer than the _small string optimization_ length of a given build environment (around 12-26 characters, depending on various factors).

```suggestion
	return std::move(ss).str();
```

Note that this is only needed here because the function returns the output of a method on `ss`. Normally the compiler can figure this out itself if directly returning a local variable (via _return value optimization_).

File: test/coverage/datamodel/UuidSerializerTest.cpp
Comment (gregmedd) [inline test/coverage/datamodel/UuidSerializerTest.cpp:65]: It might be worth adding a few more test cases here. For example, a zero-length string, a string that is the right length and has the `-` in the right place, but one character isn't valid hex (e.g. put a `z` in there)

File: test/coverage/datamodel/UuidSerializerTest.cpp
Comment (gregmedd) [inline test/coverage/datamodel/UuidSerializerTest.cpp:56]: It wouldn't hurt to test a couple more values, such as one with mixed case letters and one with some leading zeros.

File: test/coverage/datamodel/UuidSerializerTest.cpp
Comment (gregmedd) [inline test/coverage/datamodel/UuidSerializerTest.cpp:46]: It wouldn't hurt to add a few extra test values, such as one that might produce leading zeros in each of the segments.

File: test/coverage/datamodel/UuidSerializerTest.cpp
Comment (gregmedd) [inline test/coverage/datamodel/UuidSerializerTest.cpp:183]: Is this a duplicate of the minimum value test case?

**Review by gregmedd — CHANGES_REQUESTED** [review]
For the most part, this looks good. I left a comment about the string format with a link to the relevant RFC, and suggested some additional test cases. Let me know if you have any questions.

File: test/coverage/datamodel/UuidSerializerTest.cpp
Comment (BishrutSubedi) [reply test/coverage/datamodel/UuidSerializerTest.cpp:183]: yes, thanks for the catch

File: src/datamodel/serializer/Uuid.cpp
Comment (BishrutSubedi) [reply src/datamodel/serializer/Uuid.cpp:50]: Thanks, I've updated this to use htonl and ntohl. It should work for either of endianness

**Review by stevenhartley — CHANGES_REQUESTED** [review]

File: src/datamodel/serializer/Uuid.cpp
Comment (BishrutSubedi) [reply src/datamodel/serializer/Uuid.cpp:1]: Yes, sounds good. This serializer follows the 4-2-2-2-6 hextOctet format per RFC9562

The formatting part placing those bits in correct order. This could look bit different is our Uuid is custom defined, but those bits are placed correctly in serializer as we expect the uuid to look. I'm thinking this adds to readability as it is easier to see what fields are in UUID and the reason behind formatting

Comment (gregmedd) [reply src/datamodel/serializer/Uuid.cpp:1]: The variable names are version specific, but the actual conversion function looks to be version agnostic. It takes all the same bits and puts them in all the same places.

File: src/datamodel/serializer/Uuid.cpp
Comment (gregmedd) [inline src/datamodel/serializer/Uuid.cpp:129]: The more I think about it, the more I start to think we might not need this at all. Converting to/from a string makes a lot of sense for debugging or as a developer interface for loading data, but I'm having trouble coming up with an actual use case for the byte serialization.

I know there was a reason I put this in here back when I did the API, but that might have been a remnant of the way the old API worked.

We can keep it here for now and consider removing it later if there's no need for it.

File: test/coverage/datamodel/UuidSerializerTest.cpp
Comment (gregmedd) [inline test/coverage/datamodel/UuidSerializerTest.cpp:92]: The mixed case letters should be in a string that is deserialized. Otherwise this test case is perfect for the leading zeros.

**Review by gregmedd — APPROVED** [review]
I'm ok with this. There is one small note on a test case.

I don't think we should hold up the PR while we decide if we need the byte serializer.

**Review by gregmedd — CHANGES_REQUESTED** [review]
Ok, I actually have just one more change to ask for. Can you rebase this onto main (make sure it is the latest main from eclipse-uprotocol/up-cpp and not your fork) so that the CI can run?

While you are doing that, I would also appreciate it if you could use interactive rebase to squash your commits and clean up the commit messages. You don't have to squash down to one commit - whatever number clearly shows the units of work.

[This article provides some good guidance on commit messages](https://cbea.ms/git-commit/).

**Comment by BishrutSubedi** [comment]
sounds good, I will squash to 1 commit with descriptive commit msg

File: test/coverage/datamodel/UuidSerializerTest.cpp
Comment (gregmedd) [inline test/coverage/datamodel/UuidSerializerTest.cpp:115]: The mixed case letters should be in here. They don't matter in the integer literals
```suggestion
	std::string uuid_str = "00001234-5678-90ab-feDc-ba0987600000";
```

**Review by stevenhartley — APPROVED** [review]
LGTM

File: src/datamodel/serializer/Uuid.cpp
Comment (billpittman) [inline src/datamodel/serializer/Uuid.cpp:49]: NITPICK: How about a comment showing the "correct format"

// format : 12345678-1234-1234-1234-1234567890123
// index  : 012345678901234567890123456789012345

That would look better in monospace, but I can't figure out how to do it in a comment.

File: src/datamodel/serializer/Uuid.cpp
Comment (billpittman) [inline src/datamodel/serializer/Uuid.cpp:28]: Lots of shifting and bitmasking -- maybe more constexpr would help make it clearer ?  The UUID validator cleaned up amazing well, I'm hoping for the same(ish).

File: src/datamodel/serializer/Uuid.cpp
Comment (billpittman) [inline src/datamodel/serializer/Uuid.cpp:77]: NITPICK: Sorry :(  -- again maybe a comment showing how the UUID is deserialized

// format : 12345678-1234-1234-1234-1234567890123
// layout : __msb___-_msb-vcnt-____-_____RAND___
// v - version
// cnt -counter

OR whatever - I know you know this stuff now, but in 6 months we're ALL going to forget it.

File: src/datamodel/serializer/Uuid.cpp
Comment (billpittman) [inline src/datamodel/serializer/Uuid.cpp:100]: I know what you're doing, but this is a bit unclear.  Is there no bytes.data[4] way of indexing this array ?
I know these are 32-bit values being manipulated, so fixed offsets 0, 4, 8, 12 wouldn't be confusing to anyone.
I might even go so far as 
constexpr int MSB_HIGH=0
constexpr int MSB_LOW_ =4    (extra underscore to make the constants the same length so the code lines up better?)

I'm less firm about this change, so use your best judgement.

File: src/datamodel/serializer/Uuid.cpp
Comment (billpittman) [inline src/datamodel/serializer/Uuid.cpp:122]: See serialize comments.

File: test/coverage/datamodel/UuidSerializerTest.cpp
Comment (billpittman) [inline test/coverage/datamodel/UuidSerializerTest.cpp:48]: Makes me want to ask --- The code doesn't care about case .... does the spec specify lower vs upper case for hex digits?

File: test/coverage/datamodel/UuidSerializerTest.cpp
Comment (billpittman) [inline test/coverage/datamodel/UuidSerializerTest.cpp:67]: KUDOS : I don't know if you intentionally mixed the cases here - but GOOD test!

Comment (billpittman) [reply test/coverage/datamodel/UuidSerializerTest.cpp:67]: Oh you did do it on purpose (just saw the name of the test) -- well comment still applies - good test!

File: test/coverage/datamodel/UuidSerializerTest.cpp
Comment (billpittman) [inline test/coverage/datamodel/UuidSerializerTest.cpp:138]: Ok there's a missing one character test .... is there an extra character test ?  (IE too long of a string).  If there is, I must have missed it.

**Review by billpittman — CHANGES_REQUESTED** [review]
Mostly nitpicks - use your best judgement on those.

On the tests, looks great, I liked your use of sequential (expected, easy-to-identify) strings.

File: test/coverage/datamodel/UuidSerializerTest.cpp
Comment (billpittman) [inline test/coverage/datamodel/UuidSerializerTest.cpp:132]: Hmmm - for the purpose of code coverage, shouldn't this be multiple tests, one for each dash position ?
x----
-x---
--x-- etc ?

File: test/coverage/datamodel/UuidSerializerTest.cpp
Comment (billpittman) [inline test/coverage/datamodel/UuidSerializerTest.cpp:149]: Actually maybe this is the one that should be split into multiple tests -- totally up to you.

File: test/coverage/datamodel/UuidSerializerTest.cpp
Comment (BishrutSubedi) [reply test/coverage/datamodel/UuidSerializerTest.cpp:48]: there are some other cases of mixed cases which should cover this

File: test/coverage/datamodel/UuidSerializerTest.cpp
Comment (BishrutSubedi) [reply test/coverage/datamodel/UuidSerializerTest.cpp:132]: In term of coverage this should be good, but I'll add few more test for other dash placement

File: test/coverage/datamodel/UuidSerializerTest.cpp
Comment (BishrutSubedi) [reply test/coverage/datamodel/UuidSerializerTest.cpp:138]: yeah, this test case slipped my mind. I'll add it in next commit

**Comment by BishrutSubedi** [comment]
Thank you, I've updated PR with the suggested changes

**Comment by billpittman** [comment]
> Thank you, I've updated PR with the suggested changes

No thank you.  Looks great - thanks for the extra effort.  Approved!

**Review by billpittman — APPROVED** [review]

File: src/datamodel/serializer/Uuid.cpp
Comment (gregmedd) [reply src/datamodel/serializer/Uuid.cpp:1]: Since we got a LGTM from @stevenhartley, I'm going to go ahead and resolve this.

File: test/coverage/datamodel/UuidSerializerTest.cpp
Comment (gregmedd) [reply test/coverage/datamodel/UuidSerializerTest.cpp:48]: The spec is case insensitive: https://www.rfc-editor.org/rfc/rfc9562.html#name-uuid-format

**Review by gregmedd — APPROVED** [review]
Looks good to me. Can you make sure that the commits are rebased so that the conan change is dropped? It should already be committed to main. Otherwise this is ready to go.

**Pull request merged**
